File: glmm/src/mcse.c
#include "myheader.h"
/*problem with gamma and squaretop causing issue with numsum: overflow/underflow*/

/*
 gamma: scalar
 thing: scalar
 squaretop: vector length m
 numsum: vector of length (nbeta + T)^2 = npar^2
 y: vector of length n
 Umat: myq by m matrix.  one COLUMN of Umat used at a time. Umat must be the R mat transposed
 myq: total number of random effects (q). scalar. 120 for salamanders.
 m: Monte Carlo sample size. scalar.
scalar.
 x: n by nbeta matrix
 n: sample size. scalar. 
 nbeta: scalar equal to the dim of beta (number of fixed effects)
 beta: vector length nbeta
 z: n by myq matrix
 Dinvfornu: square, symmetric matrix. myq x myq. D is variance matrix.
 logdetDinvfornu: scalar equal to the log determinant for D's inverse
 family_glmm:  int scalar
 Dstarinv: square, symmetric matrix. myq x myq. Dstar is variance matrix from PQL.
 logdetDstarinv: scalar equal to the log determinant for Dstar's inverse
 ustar: vector containing PQL predictions for random effects. length = myq.
 Sigmuhinv: variance matrix for normal used in importance sampling dist. myq x myq.
 logdetSigmuhinv: scalar. 
 pee: vector of length nps. vector of sampling proportions (usually 1/3, 1/3, 1/3)
 nps: scalar
 T: scalar equal to the number of variance components
 nrandom: vector of length T. 
 meow: vector of ints. length T+1. tells us which random effects go with each variance.
 nu: vector of length T. equals the variance components.
 zeta: int equal to the df for the t distrib. (p 13 of design doc)
 tconst: scalar. (p13 of design doc, equation 59)
 ntrials: is a vec of ints with length equal to n
 lfuval: scalar
 lfyuval: scalar
 wts: weights for the observations (to calculate weighted likelihood).  vector length n.

 
*/
void mcsec(double *gamma, double *thing, double *squaretop, double *numsum, double *y, double *Umat, int *myq, int *m, double *x, int *n, int *nbeta, double *beta, double *z, double *Dinvfornu, double *logdetDinvfornu, int *family_glmm, double *Dstarinv, double *logdetDstarinv, double *ustar, double *Sigmuhinv, double *logdetSigmuhinv, double *pee, int *nps, int *T, int *nrandom, int *meow, double *nu, int *zeta, double *tconst,  int *ntrials, double *lfuval, double *lfyuval, double *wts)
{
	double *Uk = Calloc(*myq, double);
	int Uindex = 0;

	double db1 = 0.0;
	double *double1 = &db1; /*temp to hold info that will be put into lfuval*/

	/* Calculate xbeta, needed to calculate eta for each Uk=U[k,] in R notation */
	double *xbeta = Calloc(*n,double);
	matvecmult(x,beta,n,nbeta,xbeta);
	double *zu = Calloc(*n,double);
	double *eta = Calloc(*n,double);
	double *qzeros = Calloc(*myq,double);
	double tempmax = 1.0;
	double *lfutwidpieces = Calloc(*nps,double);
	double diffs = 0.0;
	/*double lfuval = 1.1;
	double lfyuval = 1.1;*/
	double lfutwid = 1.1;
	double *b = Calloc(*m,double);
	double a = 0.0;

	for(int k = 0; k < *m; k++){
		/*start by getting Uk  */
		for(int i = 0; i < *myq; i++){
			Uk[i] = Umat[Uindex];
			Uindex++;
		}

		/* calculate eta for this value of Uk
		first calculate ZUk*/
		matvecmult(z, Uk, n, myq, zu);

		/* then add xbeta+zu to get current value of eta */
		addvec(xbeta, zu, n, eta);

		/*log f_theta(u_k) goes into lfuval*/
		distRandGenC(Dinvfornu, logdetDinvfornu, myq, Uk, qzeros, lfuval); /*&*/

		/* log f_theta(y|u_k) value goes into lfyuval */
		elval(y, n, nbeta, eta, family_glmm, ntrials, wts, lfyuval); /*&*/

		/* value of log f~_theta(u_k) 
		first calculate value of log f~_theta(u_k) for 3 distribs used 
		and find largest value as you go */
		/* distRandGenC(Dinvfornu,logdetDinvfornu, myq, Uk, qzeros, double1);
		first piece is dist for t(0,Dstar)*/
/*		tempmax=*lfuval;*/
/*		lfutwidpieces[0]=*lfuval; */
		tdist(Dstarinv, myq, Uk, zeta, tconst, double1);
		lfutwidpieces[0] = *double1;
		tempmax = *double1;
		
		/*second piece is for N(ustar,Dstar)  */
		distRandGenC(Dstarinv,logdetDstarinv, myq, Uk, ustar, double1);
		lfutwidpieces[1] = *double1;
		if(*double1 > tempmax){tempmax = *double1;}

		/*third piece is for N(ustar,complicated var)  */
		distRandGenC(Sigmuhinv,logdetSigmuhinv, myq, Uk, ustar, double1);
		lfutwidpieces[2] = *double1;
		if(*double1>tempmax){tempmax = *double1;}

		lfutwid=0.0;

		/* calculate diffs= lfutwidpieces-tempmax */
		/* and lfutwid = tempmax + log(sum(pee[i]*exp(diffs))) */
		for(int i = 0; i<*nps; i++){
			diffs = lfutwidpieces[i]- tempmax;
			lfutwid+= pee[i]*exp(diffs); /*sum(pee[i]*exp(diffs))*/
		}
		lfutwid = log(lfutwid)+tempmax; /* finishes lfutwid calc */

		b[k] = *lfuval+ *lfyuval-lfutwid;

		if(k==0){a = b[k];}
		if(b[k]>a){a = b[k];}

		squaretop[k] = exp(2* *lfuval + 2* *lfyuval - 2*lfutwid); /*also exp(2*b[k])?*/
        /*tryfix[k] = b[k];*/
	}

	Free(lfutwidpieces);

	/* Calculate unnormalized log weights  */
	thing[0] = 0;
	for(int i = 0; i<*m; i++){
		thing[0]+= exp(b[i]-a);
	}
	Free(b);

	/* Calculate gamma */
	gamma[0] = exp(a) * *thing / *m;
    /*gamma[0] = exp(a - log(*m) +log(*thing));*/


/* now going to do second loop, for numerator */
	int npar = *nbeta + *T;
	double *lfugradient = Calloc(*T, double);
	double *lfuhess = Calloc((*T)*(*T), double);
	double *lfyugradient = Calloc(*nbeta, double);
	double *lfyuhess = Calloc((*nbeta)*(*nbeta), double);
	double *Gpiece = Calloc(npar, double);
	double *GGT = Calloc(npar*npar, double);

	Uindex = 0;	
	int Gindex = 0;
	int intone = 1;

	for(int k = 0; k<*m; k++){
		/*start by getting Uk  */
		for(int i = 0; i<*myq; i++){
			Uk[i] = Umat[Uindex];
			Uindex++;
		}

		/* calculate eta for this value of Uk
		first calculate ZUk*/
		matvecmult(z, Uk, n, myq, zu);

		/* then add xbeta+zu to get current value of eta */
		addvec(xbeta, zu, n, eta);

		/* calculate lfu gradient and hessian */
		distRand3C(nu, qzeros, T, nrandom, meow, Uk, lfugradient, lfuhess);

		/* calculate gradient and hessian log f_theta(y|u_k) */
		elGH(y, x, n, nbeta, eta, family_glmm, ntrials, wts, lfyugradient, lfyuhess);

		/*now I have lfugradient and lfyugradient*/

		/*create Gpiece to hold nabla log f_theta (uk,y)*/
		Gindex=0;
		for(int i = 0; i<*nbeta; i++){
			Gpiece[Gindex] = lfyugradient[i];
			Gindex++;
		}
		for(int i = 0; i<*T; i++){
			Gpiece[Gindex] = lfugradient[i];
			Gindex++;
		}
		/*now Gpiece is ready*/

		/*calculate GGT Gpiece t(Gpiece)*/
		matmatmult(Gpiece, Gpiece, &npar, &intone, &npar, GGT);
		
		/*add on GGT * weight to numsum. this corresponds to numpieces from R*/
		for(int i = 0; i < npar*npar; i++){
			numsum[i]+= GGT[i] * squaretop[k];
		}

	} /* ends  k loop */

	Free(GGT);
	Free(Gpiece);
	Free(Uk);
	Free(xbeta);
	Free(zu);
	Free(eta);
	Free(lfugradient);
	Free(lfuhess);
	Free(lfyugradient);
	Free(lfyuhess);
	Free(qzeros);



}

File: glmm/R/mcse.R
#ntrials is a vector with length equal to length(y). if Bern or Poisson, ntrials is a vec of 1s

mcvcov <- function(object){

	mod <- object
	varcomps.equal <- mod$varcomps.equal
	levs<-ordered(unique(varcomps.equal))
	

	beta <- mod$beta
	nu <-mod$nu
	umat<-mod$umat
	m<-nrow(umat)

	x <- mod$x
	y <- mod$y
	random <- mod$z
	z<-list()
	for(i in 1:length(levs)){
	    if(levs[i]!=i) stop("The numbers in the vector varcomps.equal must be consecutive. You must start at 1 and then each entry must be the next consecutive number or a repeat of a previous number.")
	    these<-varcomps.equal==i
	    thesemats<-random[these]
	    z[[i]]<-do.call(cbind,thesemats) 
	}
	names(z)<-mod$varcomps.names #z is list of length = number of varcomps
	
	Z = do.call(cbind, z)
	Tee <-length(z)
	nrand<-lapply(mod$z, ncol)
	nrandom<-unlist(nrand)


	family.glmm<-getFamily(mod$family.glmm)
	if(family.glmm$family.glmm=="bernoulli.glmm"){family_glmm=1}	
	if(family.glmm$family.glmm=="poisson.glmm"){family_glmm=2}	
	if(family.glmm$family.glmm=="binomial.glmm"){family_glmm=3}	

	pvec <-mod$pvec
	p1 <- pvec[1]
	p2 <- pvec[2]
	p3 <- pvec[3]
	zeta <- mod$zeta
	beta.pql <- mod$beta.pql
	nu.pql <- mod$nu.pql
	
	#need to get D*
	eek<-getEk(mod$z)
	Aks<-Map("*",eek,nu.pql)
	D.star<-addVecs(Aks) 
	D.star<-diag(D.star)
	D.star.inv<-solve(D.star)

	#need to recreate the variance matrix of  imp sampling distribution
	
	eta.star<-as.vector(x%*%beta.pql+Z%*%mod$u.pql)
	cdouble<-bernoulli.glmm()$cpp(eta.star) #still a vector
	cdouble<-diag(cdouble)
	Sigmuh.inv<- t(Z)%*%cdouble%*%Z+D.star.inv
	Sigmuh<-solve(Sigmuh.inv)

	Dstinvdiag<-diag(D.star.inv)

	logdet.D.star.inv<-	-sum(log(diag(D.star)))
	logdet.Sigmuh.inv<-sum(log(eigen(Sigmuh.inv,symmetric=TRUE)$values))
 	myq<-nrow(D.star.inv)

	tconst<-tconstant(zeta,myq,Dstinvdiag)

	#for the particular value of nu we're interested in, need to prep for distRandGenC
	eek<-getEk(mod$z)
	preDinvfornu<-Map("*",eek,(1/nu))
	Dinvfornu<-addVecs(preDinvfornu)
	logdetDinvfornu<-sum(log(Dinvfornu))
	Dinvfornu<-diag(Dinvfornu)
	
	meow<-rep(1,Tee+1)
	meow[1]<-0
	throwaway<-Tee+1
	meow[2:throwaway]<-cumsum(nrandom)
	
	pea<-mod$pvec
	n<-nrow(mod$x)

	nbeta<- length(beta)
	npar <- length(beta) + length(nu)
	squaretop <- rep(0,m)

	if(is.null(mod$weights)){
	  wts <- rep(1,length(y))
	} else{
	  wts <- mod$weights
	}

	# only YOU can prevent segfault errors
	# prevent segfault errors by checking dims
	stopifnot(length(y) == n)
	stopifnot(nrow(mod$x) == n)
	stopifnot(ncol(mod$x) == nbeta)
	stopifnot(length(beta) == nbeta)
	stopifnot(nrow(Z) == n)
	stopifnot(ncol(Z) == myq)
	stopifnot(nrow(Dinvfornu) == myq)
	stopifnot(ncol(Dinvfornu) == myq)
	stopifnot(length(logdetDinvfornu) == 1)
	stopifnot(length(family_glmm) == 1)
	stopifnot(nrow(D.star.inv) == myq)
	stopifnot(ncol(D.star.inv) == myq)
	stopifnot(length(logdet.D.star.inv) == 1)
    stopifnot(length(mod$u.pql) == myq)
    stopifnot(nrow(Sigmuh.inv) == myq)
    stopifnot(ncol(Sigmuh.inv) == myq)
    stopifnot(length(logdet.Sigmuh.inv) == 1)
    stopifnot(length(Tee) == 1)
    stopifnot(length(nrandom) == Tee)
    stopifnot(length(nrandom) == length(nu))
    stopifnot(length(meow) == (Tee+1))
    stopifnot(length(nu) == Tee)
    stopifnot(length(zeta) == 1)
    stopifnot(length(tconst) == 1)
    stopifnot(length(mod$mod.mcml$ntrials) == n)
    stopifnot(length(wts) == n)
    
	
	stuff<-.C(C_mcsec, 
	          as.double(0.0), # gamma: scalar
	          as.double(0.0), # thing: scalar
	          as.double(squaretop), # squaretop: vector length m
	          numsum = as.double(rep(0, npar^2)), # numsum: vector of length  npar^2
	          as.double(mod$y), # y: vector of length n
	          as.double(t(umat)), # Umat: myq by m matrix.
	          as.integer(myq), # scalar.
	          as.integer(m), # scalar.
	          as.double(mod$x), # n by nbeta matrix
	          as.integer(n), #scalar
	          as.integer(nbeta),  #scalar
	          as.double(beta), # vector length nbeta
	          as.double(Z), # n by myq matrix
	          as.double(Dinvfornu), # myq x myq
	          as.double(logdetDinvfornu), #scalar
	          as.integer(family_glmm), #scalar
	          as.double(D.star.inv), # myq x myq.
	          as.double(logdet.D.star.inv),  #scalar
	          as.double(mod$u.pql), # length = myq
	          as.double(Sigmuh.inv),  # myq x myq.
	          as.double(logdet.Sigmuh.inv), # scalar
	          pea=as.double(pea), # vector of length nps
	          nps=as.integer(length(pea)), #scalar
	          Tee=as.integer(Tee), # scalar equal to the number of variance components
	          nrandom=as.integer(nrandom), # vector of length T. 
	          meow=as.integer(meow), # vector of length T+1.
	          nu=as.double(nu), # vector of length T
	          zeta=as.integer(zeta), #scalar
	          tconst=as.double(tconst), #scalar.
	          ntrials=as.integer(mod$mod.mcml$ntrials), #vector of length n
	          as.double(0.0), # lfuval: scalar
	          as.double(0.0), #  lfyuval: scalar
	          wts=as.double(wts)) # vector length n. (to calculate weighted likelihood)

	vhatnum <- (1/m)*stuff[[4]]
	vhatdenom <- ( stuff[[1]]   )^2

	vhatvec <- vhatnum/vhatdenom
	Vhat <- matrix(vhatvec, nrow=npar)

	Uhat <- mod$loglike.hessian
	Uhatinv <- qr.solve(Uhat)

	out <- Uhatinv %*% Vhat %*% Uhatinv/m

	cf<-c(coef(object),varcomps(object))
	pnames<-names(cf)
	rownames(out) <- pnames
	colnames(out) <- pnames

	out

}

#ntrials is a vector with length equal to length(y). if Bern or Poisson, ntrials is a vec of 1s

mcse <- function(object){


	UinvVUinv <- mcvcov(object)

	MCSE <- sqrt(diag(UinvVUinv))
	MCSE

}




